File: 프로그래머스/lv3/42579. 베스트앨범/베스트앨범.java
import java.util.*;

class Solution {
    public ArrayList<Integer> solution(String[] genres, int[] plays) {
        ArrayList<Integer> answer = new ArrayList<>();
        
        HashMap<String, Integer> sortGenres = new HashMap<>();
        HashMap<String, HashMap<Integer, Integer>> sortPlays = new HashMap<>();
        for (int i = 0; i < plays.length; i++) {
            if (sortGenres.keySet().contains(genres[i])) {
                sortGenres.put(genres[i], sortGenres.get(genres[i]) + plays[i]);
            } else {
                sortGenres.put(genres[i], plays[i]);
                sortPlays.put(genres[i], new HashMap<>());
            }
            sortPlays.get(genres[i]).put(i, plays[i]);
        }
        List<Map.Entry<String, Integer>> sortedG = new LinkedList<>(sortGenres.entrySet());
        sortedG.sort(Map.Entry.comparingByValue(Comparator.reverseOrder()));
        for (Map.Entry<String, Integer> g : sortedG) {
            List<Map.Entry<Integer, Integer>> sortedP = new LinkedList<>(sortPlays.get(g.getKey()).entrySet());
            sortedP.sort(Map.Entry.comparingByValue(Comparator.reverseOrder()));
            int cnt = 0;
            for (Map.Entry<Integer, Integer> p : sortedP) {
                cnt++;
                answer.add(p.getKey());
                if (cnt == 2) break;
            }
        }
        
        
        return answer;
    }
}

File: 프로그래머스/lv3/42579. 베스트앨범/베스트앨범.cpp
#include <string>
#include <vector>
#include <map>
#include <algorithm>

using namespace std;

bool compareTotal(const pair<string, pair<int, vector<pair<int, int>>>> &a, const pair<string, pair<int, vector<pair<int, int>>>> &b) {
    return a.second.first > b.second.first;
}

bool compareEach(const pair<int, int> &a, const pair<int, int> &b) {
    return a.second > b.second;
}

vector<int> solution(vector<string> genres, vector<int> plays) {
    vector<int> answer;

    map<string, pair<int, vector<pair<int, int>>>> mapListing;
    for (int i = 0; i < genres.size(); i++) {
        mapListing[genres[i]].first += plays[i];
        mapListing[genres[i]].second.push_back(make_pair(i, plays[i]));
    }

    vector<pair<string, pair<int, vector<pair<int, int>>>>> vecListing(mapListing.begin(), mapListing.end());
    sort(vecListing.begin(), vecListing.end(), compareTotal);

    for (auto p1 : vecListing) {
        sort(p1.second.second.begin(), p1.second.second.end(), compareEach);
        int cnt = 0;
        for (pair<int, int> p2 : p1.second.second) {
            answer.push_back(p2.first);
            cnt++;
            if (cnt == 2) break;
        }
    }

    return answer;
}